lib/DependencyScan/ModuleDependencyScanner.cpp

```diff
@@ -545,243 +544,539 @@ ModuleDependencyScanner::getNamedSwiftModuleDependencyInfo(
+        reachableClangModules.insert(dependencies.begin(), dependencies.end());
+      }
+      allDiscoveredClangModules.insert(reachableClangModules.begin(),
+                                       reachableClangModules.end());
       continue;
-    findAllImportedClangModules(dep.ModuleName, cache, allClangModules,
-                                knownModules);
+    } else {
+      // We need to query the Clang dependency scanner for this module's
+      // unresolved imports
+      auto moduleDependencyInfo = cache.findKnownDependency(moduleID);
+
+      // Figure out which imports have already been resolved to module dependencies
+      llvm::StringSet<> resolvedImportIdentifiers;
+      for (const auto &resolvedDep : moduleDependencyInfo.getImportedSwiftDependencies())
+        resolvedImportIdentifiers.insert(resolvedDep.ModuleName);
+
+      for (const auto &depImport : moduleDependencyInfo.getModuleImports())
+        if (!resolvedImportIdentifiers.contains(depImport.importIdentifier)) {
+          unresolvedImports->push_back(depImport);
+          unresolvedImportIdentifiers.insert(depImport.importIdentifier);
+        }
+      for (const auto &depImport : moduleDependencyInfo.getOptionalModuleImports())
+        if (!resolvedImportIdentifiers.contains(depImport.importIdentifier)) {
+          unresolvedOptionalImports->push_back(depImport);
+          unresolvedOptionalImportIdentifiers.insert(depImport.importIdentifier);
+        }
+    }
   }
 
-  // Find all Swift overlays that this module depends on.
-  resolveSwiftOverlayDependencies(moduleID, allClangModules, cache,
-                                  swiftOverlayDependencies, directDependencies);
-
-  // Resolve the dependency info with dependency module information.
-  cache.resolveDependencyImports(moduleID, directDependencies.getArrayRef());
-  // Resolve the dependency info with Swift overlay dependency information.
-  if (!swiftOverlayDependencies.empty())
-    cache.setSwiftOverlayDependencies(moduleID,
-                                      swiftOverlayDependencies.getArrayRef());
-
-  ModuleDependencyIDSetVector result = directDependencies;
-  result.insert(swiftOverlayDependencies.begin(),
-                swiftOverlayDependencies.end());
-  return result.takeVector();
+  // Prepare the module lookup result collection
+  llvm::StringMap<std::optional<ModuleDependencyVector>> moduleLookupResult;
+  for (const auto &unresolvedIdentifier : unresolvedImportIdentifiers)
+    moduleLookupResult.insert(
+        std::make_pair(unresolvedIdentifier.getKey(), std::nullopt));
+
+  std::mutex CacheAccessLock;
+  auto scanForClangModuleDependency =
+      [this, &cache, &moduleLookupResult, &CacheAccessLock](Identifier moduleIdentifier) {
+        auto moduleName = moduleIdentifier.str();
+        {
+          std::lock_guard<std::mutex> guard(CacheAccessLock);
+          if (cache.hasDependency(moduleName, ModuleDependencyKind::Clang))
```

Comment:
You have clang module lookup here in the threads. I guess it should be fine since this is technically a `const` value for lookup only but I am a bit paranoid about the mistake can be made. Maybe do the cache lookup before `async` and don't capture `cache` at all will make it less error prone.